core/modules/loader/MWorkerList.cc

```diff
@@ -0,0 +1,423 @@
+                "neighborRight changing from valid old=" << old <<
+                " to new=" << nInfo.neighborRight->get());
+    }
+    if (old != nInfo.neighborRight->get()) {
+
+        LOGS(_log, LOG_LVL_INFO, "Worker=" << _wId <<
+                "neighborRight=" << nInfo.neighborRight->get());
+    }
+    _neighborsInfo.neighborRight->update(nInfo.neighborRight->get());
+}
+
+
+int MWorkerListItem::getKeyCount() const {
+    return _neighborsInfo.keyCount;
+}
+
+
+std::ostream& operator<<(std::ostream& os, MWorkerListItem const& item) {
+    os << "name=" << item._wId << " address=" << *item._udpAddress << " range(" << item._range << ")";
+    return os;
+}
+
+
+/// Set this worker's RIGHT neighbor to the worker described in 'item'.
+void MWorkerListItem::setRightNeighbor(MWorkerListItem::Ptr const& item) {
+    // Create a one shot to send a message to the worker.
+    // It knows it has worked when the worker sends a message back saying it
+    // has the correct right neighbor.
+    LOGS(_log, LOG_LVL_DEBUG," MWorkerListItem::setRightNeighbor");
+
+    auto oneShot = SetNeighborOneShot::create(_central,
+                                              shared_from_this(),
+                                              LoaderMsg::WORKER_RIGHT_NEIGHBOR,
+                                              item->getId(),
+                                              _neighborsInfo.neighborRight);
+    _central->addDoListItem(oneShot);
+}
+
+
+// TODO very similar to MWorkerListItem::setRightNeighbor, consider merging.
+void MWorkerListItem::setLeftNeighbor(MWorkerListItem::Ptr const& item) {
+    // Create a one shot to send a message to the worker.
+    // It knows it has worked when the worker sends a message back saying it
+    // has the correct left neighbor.
+    LOGS(_log, LOG_LVL_DEBUG,"MWorkerListItem::setLeftNeighbor");
+
+    auto oneShot = SetNeighborOneShot::create(_central,
+                                              shared_from_this(),
+                                              LoaderMsg::WORKER_LEFT_NEIGHBOR,
+                                              item->getId(),
+                                              _neighborsInfo.neighborLeft);
+
+    _central->addDoListItem(oneShot);
+}
+
+
+util::CommandTracked::Ptr MWorkerListItem::SendListToWorker::createCommand() {
+    auto item = mWorkerListItem.lock();
+    if (item == nullptr) {
+        // TODO: should mark set the removal flag for this doListItem
+        return nullptr;
+    }
+
+    struct SendListToWorkerCmd : public util::CommandTracked {
+        SendListToWorkerCmd(CentralMaster *centM_, MWorkerListItem::Ptr const& tItem_) : centM(centM_), tItem(tItem_) {}
+        void action(util::CmdData*) override {
+            LOGS(_log, LOG_LVL_DEBUG, "SendListToWorkerCmd::action");
+            centM->getWorkerList()->sendListTo(centM->getNextMsgId(),
+                    tItem->_udpAddress->ip, tItem->_udpAddress->port,
```

Comment:
I'm confused about locking: 
in `MWorkerListItem` there is 
`mutable std::mutex _mtx;  ///< protects _name, _address, _range`
I think _address actually means `_udpAddress` and `_tcpAddress` (because there is no _address member, and `_mtx` does protect `_udpAddress` in `getUdpAddress`), but there are places in this file where `_udpAddress` is not protected, like here. 
What are the locking requirements, and do they need to be fixed here?

Reply:
_udpAddress and _tcpAdddress are of class NetworkAddress, which is immutable. I think _udpAddress and _tcpAddress do not need protection from _mtx.

Names are given out by _udpAddress, so if a computer dies, if another comes along using the same udp address, it will get the same name.

Reply:
How is NetworkAddress immutable?